allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py

```diff
@@ -643,31 +635,54 @@ def _compute_validation_outputs(self,
         outputs['debug_info'] = []
         outputs['entities'] = []
         outputs['logical_form'] = []
+        outputs['answer'] = []
         for i in range(batch_size):
             # Decoding may not have terminated with any completed logical forms, if `num_steps`
             # isn't long enough (or if the model is not trained enough and gets into an
             # infinite action loop).
+            outputs['logical_form'].append([])
             if i in best_final_states:
-                best_action_indices = best_final_states[i][0].action_history[0]
-                action_strings = [action_mapping[(i, action_index)] for action_index in best_action_indices]
-                try:
-                    logical_form = world[i].get_logical_form(action_strings, add_var_function=False)
-                    self._has_logical_form(1.0)
-                except ParsingError:
-                    self._has_logical_form(0.0)
-                    logical_form = 'Error producing logical form'
-                if example_lisp_string:
-                    denotation_correct = self._executor.evaluate_logical_form(logical_form,
-                                                                              example_lisp_string[i])
-                    self._denotation_accuracy(1.0 if denotation_correct else 0.0)
-                outputs['best_action_sequence'].append(action_strings)
-                outputs['logical_form'].append(logical_form)
+                all_action_indices = [best_final_states[i][j].action_history[0] for j in
+                                      range(len(best_final_states[i]))]
+                found_denotation = False
+                for action_indices in all_action_indices:
+                    action_strings = [action_mapping[(i, action_index)] for action_index in action_indices]
+                    has_logical_form = False
+                    try:
+                        logical_form = world[i].action_sequence_to_logical_form(action_strings)
```

Comment:
Here, too, just execute the action sequence directly, instead of converting to a logical form first.  If we need to add a `world.evaluate_action_sequence(actions, targets)`, that's fine.

It'll simplify the logic below, also, as you don't need to worry about a `ParsingError`.

Reply:
But we're also measuring whether the parser is producing valid logical forms (`has_logical_form`), for which we need to make the logical forms. The value of that metric is usually `1.0`, but its nice to have that additional check.

Reply:
That should already be guaranteed by construction - we only allow actions that produce valid logical forms.  I don't think it's worth slowing down training to test something that should have been tested in a unit test.

Reply:
We'll also get `ExecutionErrors` if something was up with the action sequence, which is another place you could catch this.  This would catch two kinds of issues, which are the only ones that we should run into, anyway: (1) cases where we were loose in our type system, but have stricter execution requirements, like around lists vs. single items; and (2) cases where the action sequence didn't terminate, but we're trying to evaluate it anyway (keeping unfinished sequences at the end of beam search).

Reply:
I was referring to action sequences that got terminated early hence producing incomplete logical forms. Do you have an estimate of how much slow down this could cause?

Reply:
No, but anything you do on the CPU is time that's not utilizing the GPU for training.  And those action sequences will be caught with execution errors anyway, so there's no reason to go to a logical form.

Reply:
Or you could just instruct the beam search to not include unfinished sequences, which is probably the right thing to do, anyway...

Reply:
Okay, removing the `has_logical_form` metric.

Reply:
Hmm.. one more thing. We do need to show the logical form in the demo, and to be able to convert the action sequence into a logical form, we need the corresponding world objects. I think this is the best place to make a logical form and store it in the output dict. Also, this method is only called during validation, so should not significantly slow down training.

Reply:
Hmm, yeah, fair enough, that's a good point.  You could pass the world object in the output dictionary, and create the logical form in `decode`, but it's probably fine to just put it here during validation.